src/mlpack/methods/ann/layer/read_memory.hpp

```diff
@@ -0,0 +1,276 @@
+/**
+ * @file read_memory.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Definition of Read Memory layer used in Neural Turing Machine.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_READ_MEMORY_HPP
+#define MLPACK_METHODS_ANN_LAYER_READ_MEMORY_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "../visitor/delta_visitor.hpp"
+#include "../visitor/output_parameter_visitor.hpp"
+
+#include "layer_types.hpp"
+
+#include "memory_head.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * An implementation of a Read Memory used in NTM.
+ * The layer takes the controller output as input and computes the read memory
+ * content.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ReadMemory
+{
+ public:
+  /**
+   * Create the Read Memory layer object using the specified parameters.
+   *
+   * @param inSize The number of input units.
+   * @param numMem Number of memory locations.
+   * @param memSize Memory size in each memory block.
+   * @param shiftSize Circular convolutional shift size used.
+   */
+  ReadMemory(const size_t inSize,
+             const size_t numMem,
+             const size_t memSize,
+             const size_t shiftSize);
+
+  /**
+   * Feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward given the current memory content.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   * @param memory Current memory content.
+   */
+  template<typename eT>
+  void ForwardWithMemory(arma::Mat<eT>&& input,
+                         const arma::Mat<eT>&& memory,
+                         arma::Mat<eT>&& output);
+
+  /**
+   * Feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward given the current memory content.
+   *
+   * This function is used in testing the layer with MemoryTest layer.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   * @param memory Current memory content.
+   */
+  template<typename eT>
+  void ForwardWithMemoryTest(arma::Mat<eT>&& input,
+                             const arma::Mat<eT>&& memory,
+                             arma::Mat<eT>&& output)
```

Comment:
I agree also---probably the internal implementation is better.  I think it will simplify the code too; probably visitors won't be necessary to use internally.  You could just do the matrix operations directly.  I think this will be a pretty significant change, so I think if you'll redo the implementation internally, I'll save further comments for once that's done.